### PR TITLE
PERF: lazify IO imports

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/__init__.py [hunk 1]
```diff
@@ -184,6 +147,89 @@
...
+    elif key == "json_normalize":
+        from pandas.io.json._normalize import json_normalize
+
+        return json_normalize
+    raise AttributeError(f"module 'pandas' has no attribute '{key}'")
+
+
+def __dir__() -> list[str]:
+    # include lazy imports defined in __getattr__ in dir()
+    base = list(globals().keys())
```
Comment (twoertwein) [inline pandas/__init__.py:199]: Could this be simplified to `return __all__`?

Comment (jbrockmendel) [reply pandas/__init__.py:203]: it could, but i think that would render the namespace tests redundant

**Comment by MarcoGorelli** [comment]
generally in favour - do you have an estimate of how much this helps?

**Comment by jbrockmendel** [comment]
>  do you have an estimate of how much this helps?

The measurements with `python -X importtime -c "import pandas as pd"` are extremely noisy.  I eyeball it at roughly 511ms in main 472ms in this branch.

The next big chunk available would be to delay core.groupby import, which is 86ms.  Doing that requires some yak-shaving bc core.window imports it, so that needs to be made lazy (and docstrings moved xref #51823).  Also dask references pd.core.groupby.groupby.something_i_forget_what at import-time to do a docstring pinning, so we'd need to do something about that.

Following that is some smaller-but-easier stuff: stdlib imports like shutil and stuff in pd.compat.compressors that are relatively expensive, dateutil (xref #52659).

The big ones are pyarrow (87ms) and numpy (86ms).  making numpy import lazy is probably not worth the effort (though asking numpy to lazify parts of its own imports might be worthwhile).  I'd like to see the pyarrow import made lazy, especially if we don't make it required.

File: pandas/io/parsers/c_parser_wrapper.py [hunk 2]
```diff
@@ -64,10 +64,7 @@ def __init__(self, src: ReadCsvBuffer[str], **kwds) -> None:
         self.low_memory = kwds.pop("low_memory", False)
 
         # #2442
-        # error: Cannot determine type of 'index_col'
-        kwds["allow_leading_cols"] = (
-            self.index_col is not False  # type: ignore[has-type]
-        )
+        kwds["allow_leading_cols"] = self.index_col is not False
```
Comment (MarcoGorelli) [inline pandas/io/parsers/c_parser_wrapper.py:67]: why does lazifying imports change how things are type-checked here?

is it that it than `mypy` can't follow imports, and so we end up with a lot more `Any`s?

If so, then there is a tradeoff to consider...slightly faster `import pandas`, but fewer type hints whilst developing?

Comment (jbrockmendel) [reply pandas/io/parsers/c_parser_wrapper.py:67]: > why does lazifying imports change how things are type-checked here?

hmm i have no idea.  cc @simonjayhawkins ?

Comment (jbrockmendel) [reply pandas/io/parsers/c_parser_wrapper.py:67]: any idea @dr-irv?

Comment (Dr-Irv) [reply pandas/io/parsers/c_parser_wrapper.py:67]: This is some mypy funkyness.

If you insert the line `reveal_type(self.index_col)` in main, you get:
```text
pandas\io\parsers\c_parser_wrapper.py:68: error: Cannot determine type of "index_col"  [has-type]
pandas\io\parsers\c_parser_wrapper.py:68: note: Revealed type is "Any"
```
With this PR, the first message about "Cannot determine type" goes away.  But it still sees the type of `self.index_col` as `Any`.  So I guess this is a good thing

**Comment by jbrockmendel** [comment]
No clear interest here, closing.

**Pull request closed without merging**